sklearn/feature_selection/mutual_info.py

```diff
@@ -0,0 +1,375 @@
+    ----------
+    .. [1] B. C. Ross "Mutual Information between Discrete and Continuous
+       Data Sets". PLoS ONE 9(2), 2014.
+    """
+    c = c.reshape((-1, 1))
+    n_samples = c.size
+
+    nn = NearestNeighbors(n_neighbors=n_neighbors)
+
+    radius = np.empty(n_samples)
+    label_counts = np.empty(n_samples)
+    for label in np.unique(d):
+        mask = d == label
+        count = np.sum(mask)
+        if count > 1:
+            nn.set_params(n_neighbors=min(n_neighbors, count - 1))
+            nn.fit(c[mask])
+            r = nn.kneighbors()[0]
+            radius[mask] = np.nextafter(r[:, -1], 0)
+        else:
+            radius[mask] = 0
+        label_counts[mask] = count
+
+    nn.set_params(algorithm='kd_tree')
+    nn.fit(c)
+    ind = nn.radius_neighbors(radius=radius, return_distance=False)
+    neighbor_counts = np.array([i.size for i in ind])
+
+    mi = (digamma(n_samples) + digamma(n_neighbors) -
+          np.mean(digamma(label_counts)) -
+          np.mean(digamma(neighbor_counts + 1)))
+
+    return max(0, mi)
+
+
+def _compute_mi(x, y, x_categorical, y_categorical, n_neighbors=3):
+    """Compute mutual information between two variables.
+
+    This is a simple wrapper which selects a proper function to call based on
+    whether `x` and `y` are categorical or not.
+    """
+    if x_categorical and y_categorical:
+        return mutual_info_score(x, y)
+    elif x_categorical and not y_categorical:
+        return _compute_mi_cd(y, x, n_neighbors)
+    elif not x_categorical and y_categorical:
+        return _compute_mi_cd(x, y, n_neighbors)
+    else:
+        return _compute_mi_cc(x, y, n_neighbors)
+
+
+def _iterate_csc_columns(X, columns=None):
+    """This little hack returns a densified columns when iterating over a
+    sparse matrix, instead of constructing a sparse matrix for every column
+    that is expensive.
+
+    Parameters
+    ----------
+    X : csc_matrix, shape (n_samples, n_features)
+        Matrix over which to iterate.
+    columns : iterable or None, default None
+        Indices of columns to iterate over. If None, iterate over all columns.
+
+    Yields
+    ------
+    xi : ndarray, shape (n_samples,)
+        Columns of `X` in dense format.
+    """
+    n_features = X.shape[1]
+    X_indices = X.indices
+    X_data = X.data
+    X_indptr = X.indptr
+
+    if columns is None:
+        columns = range(n_features)
+
+    for i in columns:
+        column = np.zeros(X.shape[0])
+        startptr, endptr = X_indptr[i], X_indptr[i + 1]
+        nonzero_indices = X_indices[startptr:endptr]
+        column[nonzero_indices] = X_data[startptr:endptr]
+        yield column
+
+
+def _iterate_dense_columns(X, columns=None):
+    """Simple convenience iterator over ndarray columns.
+
+    Works the same as _iterate_csc_columns, but for dense ndarray.
+    """
+    if columns is None:
+        columns = range(X.shape[1])
+
+    for i in columns:
+        yield X[:, i]
+
+
+def _get_column(X, i):
```

Comment:
You could now do this inside `_iterate_csc_columns` and remove code duplication,